### PR TITLE
Add an example for memory-efficient techniques for recommendation models

### PR DESCRIPTION
**Pull request by ksalama**
Hello @fchollet -

I created this example to  demonstrates two techniques for building memory-efficient recommendation models
by reducing the size of the embedding tables, without sacrificing the model effectiveness:
1. [Qutient-remainder trick](https://arxiv.org/abs/1909.02107) aims to reduce the number of embedding vectors to store, yet produces unique embedding vector for each item without explicit definition.
2. [Mixed Dimension embeddings](https://arxiv.org/abs/1909.11810) aims to store embedding vectors with mixed dimensions, where less popular items have reduced dimension embeddings.

I put it under `keras_recipes`, but it can also go under `structured_data` if you prefer.

Cheers

### CODE REVIEW

File: examples/keras_recipes/memory_efficient_embeddings.py [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+        optimizer=keras.optimizers.Adam(learning_rate),
+        loss=tf.keras.losses.MeanSquaredError(),
+        metrics=[keras.metrics.MeanAbsoluteError(name="mae")],
+    )
+    # Read the training data.
+    train_dataset = get_dataset_from_csv("train_data.csv", batch_size)
+    # Read the test data.
+    eval_dataset = get_dataset_from_csv("eval_data.csv", batch_size, shuffle=False)
+    # Fit the model with the training data.
+    history = model.fit(train_dataset, epochs=num_epochs, validation_data=eval_dataset)
```
Comment (fchollet) [inline examples/keras_recipes/memory_efficient_embeddings.py:123]: You don't know in advance which epoch will perform best, so use a ModelCheckpoint and EarlyStopping callback (with a large number of epochs) to stop training after the model starts overfitting, then recover the best model

File: examples/keras_recipes/memory_efficient_embeddings.py [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+
+    # Evaluate the model on the test data.
+    _, rmse = model.evaluate(eval_dataset, verbose=0)
```
Comment (fchollet) [inline examples/keras_recipes/memory_efficient_embeddings.py:126]: Prefer evaluating on a separate test set at the end (after reloading the best model). If you don't want to use a separate test set then you can skip this step (the model is already evaluated during training)

File: examples/keras_recipes/memory_efficient_embeddings.py [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+
+"""
+### Implement embedding encoder
+"""
+
+
+def embedding_encoder(vocabulary, embedding_dim, num_oov_indices=0, name=None):
+    return keras.Sequential(
+        [
+            StringLookup(
```
Comment (fchollet) [inline examples/keras_recipes/memory_efficient_embeddings.py:141]: On GPU/TPU you will want to put the StringLookup layer in the tf.data pipeline, not in the model (and only migrate it back to the model to export an inference model). Though on CPU having StringLookup in the model is fine

File: examples/keras_recipes/memory_efficient_embeddings.py [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+            layers.Embedding(
+                input_dim=len(vocabulary) + num_oov_indices, output_dim=embedding_dim
+            ),
+        ],
+        name=f"{name}_embedding" if name else None,
+    )
+
+
+"""
+### Implement the standard model
```
Comment (fchollet) [inline examples/keras_recipes/memory_efficient_embeddings.py:153]: "baseline model"

File: examples/keras_recipes/memory_efficient_embeddings.py [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+"""
+
+
+def create_standard_model():
```
Comment (fchollet) [inline examples/keras_recipes/memory_efficient_embeddings.py:157]: baseline_model

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! It looks great, I think it will make for a very valuable and practical example.

File: examples/keras_recipes/memory_efficient_embeddings.py (hunk 3 above)
Comment (fchollet) [reply examples/keras_recipes/memory_efficient_embeddings.py:141]: To be clear, I'm not sure whether this matters in this case. I'd recommend benchmarking the two ways on the Colab GPU. If putting the layer in the tf.data pipeline doesn't provide a speedup, then let's keep it in the model.

File: examples/keras_recipes/memory_efficient_embeddings.py (hunk 5 above)
Comment (ksalama) [reply examples/keras_recipes/memory_efficient_embeddings.py:157]: Done

File: examples/keras_recipes/memory_efficient_embeddings.py (hunk 4 above)
Comment (ksalama) [reply examples/keras_recipes/memory_efficient_embeddings.py:153]: Done

File: examples/keras_recipes/memory_efficient_embeddings.py (hunk 2 above)
Comment (ksalama) [reply examples/keras_recipes/memory_efficient_embeddings.py:126]: Done

File: examples/keras_recipes/memory_efficient_embeddings.py (hunk 3 above)
Comment (ksalama) [reply examples/keras_recipes/memory_efficient_embeddings.py:141]: Ok, I tried it on a GPU Colab and putting StringLookup in the tf.data pipeline didn't provide a significant speendup (1ms difference per step!).

Maybe with a bit of optimization of the tf.data pipeline it could, however I would keep the model as is.

**Comment by ksalama** [comment]
@fchollet Thanks for your comments. I made the required changes.

File: examples/keras_recipes/memory_efficient_embeddings.py [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+    checkpoint_callback = keras.callbacks.ModelCheckpoint(
+        checkpoint_filepath, monitor="val_loss", save_best_only=True
+    )
+
+    # Compile the model.
+    model.compile(
+        optimizer=keras.optimizers.Adam(learning_rate),
+        loss=tf.keras.losses.MeanSquaredError(),
+        metrics=[keras.metrics.MeanAbsoluteError(name="mae")],
+        callbacks=[checkpoint_callback],
```
[the remark this answers is not in the document]
Comment (ksalama) [reply examples/keras_recipes/memory_efficient_embeddings.py:123]: We are not using early stopping callback.

You are suggesting **not** to use it, right?

File: examples/keras_recipes/memory_efficient_embeddings.py [hunk 7]
```diff
@@ -0,0 +1,425 @@
...
+        num_epochs=1,
+        header=False,
+        field_delim="|",
+        shuffle=shuffle,
+    )
+
+
+def run_experiment(model):
+
+    checkpoint_filepath = "/tmp/checkpoint"
```
[the remark this answers is not in the document]
Comment (ksalama) [reply examples/keras_recipes/memory_efficient_embeddings.py:113]: Done

File: examples/keras_recipes/memory_efficient_embeddings.py [hunk 8]
```diff
@@ -0,0 +1,425 @@
...
+"""
+### Implement Quotient-Remainder embedding as a layer
+
+Quotient-Remainder technique works as follows. For a set of vocabulary and  embedding size
+`embedding_dim`, instead of creating a `vocabulary_size X embedding_dim` embedding table,
+we create *two* `num_buckets X embedding_dim` embedding tables, where `num_buckets`
+is much smaller than `vocabulary_size`.
+An embedding for a given item `index` is generated via the following steps:
+
+1. Compute the quotient `q_index` as `index // num_buckets`.
```
[the remark this answers is not in the document]
Comment (ksalama) [reply examples/keras_recipes/memory_efficient_embeddings.py:219]: Done

**Comment by ksalama** [comment]
@fchollet I have address your comments. Cheers!

**Comment by ksalama** [comment]
I assume that the `Continuous integration / docker-image (pull_request)` check failure is nothing to do with my code, right?

**Review by fchollet — COMMENTED** [review]
Looking great! Please add the generated files.

**Comment by ksalama** [comment]
@fchollet - Thank you for your review. I have added the generated files.

**Review by fchollet — APPROVED** [review]
Thanks for the great contribution! 👍
